mira/dkg/web_client.py

```diff
@@ -1,104 +1,214 @@
+    endpoint: str,
+    method: Literal["get", "post"],
+    query_json: Optional[Dict[str, Any]] = None,  # Required for post
+    api_url: Optional[str] = None,
+) -> Union[List[Dict[str, Any]], Dict[str, Any], None]:
+    """A wrapper for sending requests to the REST API and returning the results
+
+    Parameters
+    ----------
+    endpoint :
+        The endpoint to send the request to.
+    method :
+        Which method to use. Must be one of 'post' and 'get'.
+    query_json :
+        The data to send with the request. Must be filled if method is 'post'.
+    api_url :
+        Provide the base URL to the REST API. Use this argument to override
+        the default set in MIRA_REST_URL or rest_url from the config file.
+
+    Returns
+    -------
+    :
+        The data sent back from the endpoint as a json, unless the response
+        is empty, in which case None is returned.
+    """
     base_url = api_url or os.environ.get("MIRA_REST_URL") or pystow.get_config("mira", "rest_url")
 
     if not base_url:
         raise ValueError(
-            "The base url for the rest api needs to either be set in the "
+            "The base url for the REST API needs to either be set in the "
             "environment using the variable 'MIRA_REST_URL', be set in the "
             "pystow config 'mira'->'rest_url' or by passing it the 'api_url' "
             "parameter to this function."
         )
 
+    # Clean base url and endpoint
     base_url = base_url.rstrip("/") + "/api" if not base_url.endswith("/api") else base_url
-
+    endpoint = endpoint if endpoint.startswith("/") else "/" + endpoint
     endpoint_url = base_url + endpoint
-    res = requests.post(endpoint_url, json=query_json)
+
+    if method == "post":
+        if query_json is None:
+            raise ValueError(f"POST request to endpoint {endpoint} requires query data")
+        res = requests.post(endpoint_url, json=query_json)
+    elif method == "get":
+        # Add query_json as params if present
+        kw = dict() if query_json is None else {"params": query_json}
+        res = requests.get(endpoint_url, **kw)
+    else:
+        raise ValueError("Method must be one of 'get' and 'post'")
+
     res.raise_for_status()
 
-    return res.json()
+    if res.text and res.json():
+        return res.json()
 
 
 def get_relations_web(
-    source_type: Optional[str] = None,
-    source_curie: Optional[str] = None,
-    target_type: Optional[str] = None,
-    target_curie: Optional[str] = None,
-    # "relations" is "relation_type" in client.query_relations
-    relations: Union[None, str, List[str]] = None,
-    relation_direction: Literal["right", "left", "both"] = "right",
-    relation_min_hops: int = 1,
-    relation_max_hops: int = 1,
-    limit: Optional[int] = None,
-    full: bool = False,
-    distinct: bool = False,
-    api_url: str = None,
-):
-    """A wrapper that call the rest API's get_relations endpoint
+    relations_model: api.RelationQuery,
```

Comment:
I just noticed the use of the api's model here as well, I guess it's more justified here given how complicated the inputs are.

Reply:
Here I instead added a code example in the docstring so anyone using the function gets explicit instruction of how to provide the RelationQuery model. See https://github.com/indralab/mira/pull/38/commits/69dd967a107050560a6431a09a7fb8c80cc972ac